File: table_statistics.py
#Используемые библиотеки
import yaml
import csv
from collections import defaultdict
import os
from dotenv import load_dotenv
#Чтение перменных из файла .env
path = r'd:\CodeRun\table_statistics\table_statistics\variables.env'
load_dotenv(path)
def load_yaml(file_path):
    '''Функция для чтения .yaml'''
    with open(file_path, 'r') as file:
        return yaml.safe_load(file)
def load_logs(file_path):
    '''Функция для чтения .csv'''
    logs = []
    with open(file_path, 'r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            logs.append(row['node_id'])
    return logs
def find_statistics_nodes(folders):
    '''Функция для поиска по узлу'''
    #Мноежство для хранения результатов
    statistics_nodes = set()
    #Функция для проверки текущей ссылки на link
    def traverse(node):
        if 'link' in node:
            return
        if node.get('name') == 'statistics':
            statistics_nodes.add(node)
            return
        #Рекурсивный обход дочерних узлов
        for key, child in node.get('values', {}).items():
            traverse(child)
    #Рекурсивный обход узлов
    for key, node in folders.items():
        traverse(node)
    return statistics_nodes
def get_all_node_ids(node):
    '''Функция для получения id узла'''
    node_ids = set()
    #Функция для проверки отношения узла с link
    def collect_ids(node):
        if 'link' in node:
            return
        if 'values' in node:
            #Рекурсивный обход дочерних ключей
            for key, child in node['values'].items():
                node_ids.add(key)
                collect_ids(child)
    collect_ids(node)
    return node_ids
def main():
    # Считывание нахождения в директории
    file_path_yaml = os.getenv("file_path_yaml")
    file_path_csv = os.getenv("file_path_csv")
    folders = load_yaml(file_path_yaml)
    logs = load_logs(file_path_csv)
    #Идентификаторы узлов statistics
    home_statistics_id = '9cae29'
    tmp_statistics_id = '82cc0d'
    def get_all_node_ids(node):
        """Рекурсивно собирает все ID узлов в поддереве"""
        node_ids = set()
        if 'values' in node:
            for key, child in node['values'].items():
                node_ids.add(key)
                node_ids.update(get_all_node_ids(child))
        return node_ids
    def find_statistics_node(folders, target_id):
        """Находит узел statistics по его ID"""
        for key, node in folders.items():
            if key == target_id:
                return node
            if 'values' in node:
                found = find_statistics_node(node['values'], target_id)
                if found:
                    return found
        return None
    #Находим узел /home/statistics
    statistics_node = find_statistics_node(folders, home_statistics_id)
    if not statistics_node:
        print("Узел 'statistics' не найден в структуре.")
    else:
        #Собираем все ID узлов в поддереве statistics
        target_ids = get_all_node_ids(statistics_node)
        target_ids.add(home_statistics_id)  # Добавляем сам узел statistics
        #Подсчет обращений
        count = sum(1 for node_id in logs if node_id in target_ids)
        print(f"Количество обращений к /home/statistics и подузлам: {count}")
#Конструкция для скрипта
if __name__ == "__main__":
    main()

File: table_statistics.java
//Вызов библиотек
import java.io.*;
import java.util.*;
import java.nio.file.Paths;
import java.nio.file.Files;
//Используемые библиотеки
public class TableStatistics {
    //Функция для чтения .yaml
    public static Map<String, Object> loadYaml(String filePath) throws IOException {
        /**
         *Функция для чтения .yaml
         *@param filePath путь к файлу YAML
         *@return Map с данными из YAML
         */
        Map<String, Object> data = new HashMap<>();
        return data;
    }
    //Функция для чтения .csv
    public static List<String> loadLogs(String filePath) throws IOException {
        /**
         *Функция для чтения .csv
         *@param filePath путь к CSV файлу
         *@return List с node_id из логов
         */
        List<String> logs = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line = reader.readLine(); //Читаем заголовок
            if (line != null && line.contains("node_id")) {
                while ((line = reader.readLine()) != null) {
                    String[] columns = line.split(",");
                    if (columns.length > 0) {
                        logs.add(columns[0]); //Предполагаем, что node_id в первом столбце
                    }
                }
            }
        }
        return logs;
    }
    
    //Функция для поиска узла statistics по ID
    public static Node findStatisticsNode(Map<String, Node> folders, String targetId) {
        for (Map.Entry<String, Node> entry : folders.entrySet()) {
            if (entry.getKey().equals(targetId)) {
                return entry.getValue();
            }
            if (entry.getValue().getValues() != null) {
                Node found = findStatisticsNode(entry.getValue().getValues(), targetId);
                if (found != null) {
                    return found;
                }
            }
        }
        return null;
    }
    
    //Функция для получения всех ID узлов в поддереве
    public static Set<String> getAllNodeIds(Node node) {
        Set<String> nodeIds = new HashSet<>();
        collectIds(node, nodeIds);
        return nodeIds;
    }
    
    private static void collectIds(Node node, Set<String> nodeIds) {
        if (node == null) return;
        if (node.getValues() != null) {
            for (Map.Entry<String, Node> entry : node.getValues().entrySet()) {
                nodeIds.add(entry.getKey());
                collectIds(entry.getValue(), nodeIds);
            }
        }
    }
    
    //Функция для преобразования Map<String, Object> в Map<String, Node>
    @SuppressWarnings("unchecked")
    private static Map<String, Node> convertToNodeMap(Map<String, Object> rawData) {
        Map<String, Node> result = new HashMap<>();
        for (Map.Entry<String, Object> entry : rawData.entrySet()) {
            if (entry.getValue() instanceof Map) {
                Node node = new Node();
                Map<String, Object> nodeData = (Map<String, Object>) entry.getValue();
                
                if (nodeData.containsKey("name")) {
                    node.setName(nodeData.get("name").toString());
                }
                if (nodeData.containsKey("link")) {
                    node.setLink(nodeData.get("link").toString());
                }
                if (nodeData.containsKey("values")) {
                    Map<String, Object> valuesData = (Map<String, Object>) nodeData.get("values");
                    node.setValues(convertToNodeMap(valuesData));
                }
                
                result.put(entry.getKey(), node);
            }
        }
        return result;
    }
    
    public static void main(String[] args) {
        try {
            //Загрузка переменных из .env файла
            String envPath = "d:\\CodeRun\\table_statistics\\table_statistics\\variables.env";
            Map<String, String> envVars = new HashMap<>();
            
            try (BufferedReader reader = new BufferedReader(new FileReader(envPath))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.contains("=") && !line.trim().startsWith("#")) {
                        String[] parts = line.split("=", 2);
                        if (parts.length == 2) {
                            envVars.put(parts[0].trim(), parts[1].trim());
                        }
                    }
                }
            }
            
            String filePathYaml = envVars.get("file_path_yaml");
            String filePathCsv = envVars.get("file_path_csv");
            
            if (filePathYaml == null || filePathCsv == null) {
                System.err.println("Не найдены необходимые переменные в .env файле");
                return;
            }
            
            //Идентификаторы узлов statistics
            String homeStatisticsId = "9cae29";
            
            //Загрузка данных
            Map<String, Object> rawData = loadYaml(filePathYaml);
            Map<String, Node> folders = convertToNodeMap(rawData);
            List<String> logs = loadLogs(filePathCsv);
            
            //Находим узел /home/statistics
            Node statisticsNode = findStatisticsNode(folders, homeStatisticsId);
            
            